crates/sol-types/src/types/calls.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{Panic, Result, Revert, SolError};
+use alloc::vec::Vec;
+
+/// A collection of ABI-encoded call-like types. This currently includes
+/// [`SolCall`] and [`SolError`].
+///
+/// [`SolCall`]: crate::SolCall
+/// [`SolError`]: crate::SolError
+///
+/// ### Implementer's Guide
+///
+/// We do not recommend implementing this trait directly. Instead, we recommend
+/// using the [`sol`][crate::sol] proc macro to parse a Solidity contract
+/// definition.
+pub trait SolCalls: Sized {
+    /// The name of this type.
+    const NAME: &'static str;
+
+    /// The minimum length of the data for this type.
+    ///
+    /// This does *not* include the selector's length (4).
+    const MIN_DATA_LENGTH: usize;
+
+    /// The number of variants.
+    const COUNT: usize;
+
+    /// The selector of this type.
+    fn selector(&self) -> [u8; 4];
+
+    /// Checks if the given selector is known to this type.
+    fn type_check(selector: [u8; 4]) -> Result<()>;
+
+    /// ABI-decodes the given data into one of the variants of `self`.
+    fn decode_raw(selector: [u8; 4], data: &[u8], validate: bool) -> Result<Self>;
+
+    /// The size of the encoded data, *without* any selectors.
+    fn encoded_size(&self) -> usize;
+
+    /// ABI-encodes `self` into the given buffer, *without* any selectors.
+    fn encode_raw(&self, out: &mut Vec<u8>);
+
+    /// ABI-encodes `self` into the given buffer.
+    #[inline]
+    fn encode(&self) -> Vec<u8> {
+        let mut out = Vec::with_capacity(4 + self.encoded_size());
+        out.extend(self.selector());
+        self.encode_raw(&mut out);
+        out
+    }
+
+    /// ABI-decodes the given data into one of the variants of `self`.
+    #[inline]
+    fn decode(data: &[u8], validate: bool) -> Result<Self> {
+        if data.len() < Self::MIN_DATA_LENGTH + 4 {
+            Err(crate::Error::type_check_fail(data, Self::NAME))
+        } else {
+            let (selector, data) = crate::impl_core::split_array_ref(data);
+            Self::decode_raw(*selector, data, validate)
+        }
+    }
+}
+
+/// A generic contract error.
+///
+/// Contains a [`Revert`] or [`Panic`] error, or a custom error.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ContractError<T> {
+    /// A contract's custom error.
+    CustomError(T),
+    /// A generic revert. See [`Revert`] for more information.
+    Revert(Revert),
+    /// A panic. See [`Panic`] for more information.
+    Panic(Panic),
+}
```

Comment:
Should there be an empty variant?

Reply:
Makes sense

Reply:
This just kinda breaks the entire thing, since now we can't assume that data has a selector. I'm not sure how to do this, I think it's better we don't include this

Reply:
yeah, I guess the question is whether we want this trait/enum to capture:
1. ALL inputs and outputs (known ABI & unknown ABI & Empty)
2. NON-EMPTY inputs and outputs (known ABI & unknown ABI)
3. KNOWN inputs and outputs (known ABI & Empty)
4. KNOWN & NON-EMPTY inputs and outputs (known ABI only)

Right now we have option 4. Do we have a plan for capturing the remainder?

Reply:
I'm thinking 4 is fine for now, maybe 3 in the future. I don't want to overload this PR